src/TreeTransform.jl

```diff
@@ -1,5 +1,342 @@
+            if newnode !== node
+                any_changes = true
+            end
+            @devassert node in keys(ctx.fixed_points)
+            @devassert newnode in keys(ctx.fixed_points)
+        end
+    end
+
+    result = ctx.fixed_points[data]
+    result
+end
+
+# Count the number of nodes reachable from the root node.
+function count_reachable_nodes(root::T) where { T }
+    counted = Set{Any}()
+
+    # Note that this is recursive.  If that becomes a problem, we can
+    # rewrite it iteratively.
+    function count(node::T) where { T }
+        node in counted && return
+        push!(counted, node)
+        for child in enumerate_children(node)
+            count(child)
+        end
+    end
+
+    count(root)
+    length(counted)
+end
+
+# Enumerate the children of the given node.
+function enumerate_children(node::T) where { T }
+    names = fieldnames(T)
+    num_fields = length(names)
+    vs = Vector{Any}(undef, num_fields)
+    for i in 1:num_fields
+        @inbounds v = getfield(node, i)
+        @inbounds vs[i] = v
+    end
+
+    vs
+end
+
+function enumerate_children(node::AbstractVector{T}) where { T }
+    node
+end
+
+function rebuild_node(ctx::RewriteContext, node::T) where { T }
+    names = fieldnames(T)
+    num_fields = length(names)
+
+    # Use a static vector to avoid allocations. Since this method is specialized on T, this
+    # code compiles into just copying the fields into stack/registers.
+    ws = StaticArrays.SizedVector{num_fields,Any}(nothing for i in 1:num_fields)
+
+    # True if all the fields remain `===`.
+    all_same = true
+
+    for i in 1:num_fields
+        @inbounds v = getfield(node, i)
+        w = fixed_point(ctx, v, true)
+        all_same &= v === w
+        @inbounds ws[i] = w
+    end
+
+    if all_same
+        return node
+    else
+        return T(ws...)
+    end
+end
+
+function rebuild_node(ctx::RewriteContext, node::Expr)
+    head = fixed_point(ctx, node.head, true)
+    args = fixed_point(ctx, node.args, true)
+    if (head === node.head) && (args === node.args)
+        return node
+    else
+        return Expr(node.head, args...)
+    end
+end
+
+function rebuild_node(ctx::RewriteContext, node::AbstractArray{T}) where { T }
+    length(node) == 0 && return node
+    rewritten = Base.Broadcast.broadcast(ctx.fixed_point, node)
+    for i in eachindex(node)
+        @inbounds node[i] !== rewritten[i] && return rewritten
+    end
+    return node
+end
+
+#
+# Transform a node by applying a single iteration of the transformation function
+#
+function xform(ctx::RewriteContext, @nospecialize(data))
+    ctx.transformation_count += 1
+    if ctx.max_transformations_per_node > 0 && ctx.transformation_count > 100
+        if ctx.max_transformations == 0
+            # compute the maximum number of transformations
+            ctx.max_transformations =
+                ctx.max_transformations_per_node * count_reachable_nodes(ctx.root_node)
```

Comment:
I think this happens only once, right?

Reply:
Yes, due to the tests in lines 282 and 283.